analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.fasten.analyzer.qualityanalyzer;
+
+import eu.fasten.analyzer.qualityanalyzer.data.QAConstants;
+
+import eu.fasten.core.plugins.KafkaPlugin;
+import eu.fasten.core.plugins.DBConnector;
+
+import org.json.JSONObject;
+import org.pf4j.Extension;
+import org.pf4j.Plugin;
+import org.pf4j.PluginWrapper;
+
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.List;
+import java.util.Collections;
+
+import org.jooq.DSLContext;
+
+
+public class QualityAnalyzerPlugin extends Plugin {
+
+    public QualityAnalyzerPlugin(PluginWrapper wrapper) {
+        super(wrapper);
+    }
+
+    @Extension
+    public static class QualityAnalyzer implements KafkaPlugin, DBConnector {
+
+        private final Logger logger = LoggerFactory.getLogger(QualityAnalyzer.class.getName());
+        private String consumerTopic = "fasten.RapidPlugin.out";
+        private MetadataUtils utils = null;
+
+        @Override
+        public void setDBConnection(Map<String, DSLContext> dslContexts) {
+            this.utils = new MetadataUtils(dslContexts);
+        }
+
+        @Override
+        public Optional<List<String>> consumeTopic() {
+            return Optional.of(Collections.singletonList(consumerTopic));
+        }
+
+        @Override
+        public void setTopic(String topicName) {
+            this.consumerTopic = topicName;
+        }
+
+        @Override
+        public void consume(String kafkaMessage) {
+
+            logger.info("Consumed: " + kafkaMessage);
+
+            var jsonRecord = new JSONObject(kafkaMessage);
+            String forge = null;
+
+            if (jsonRecord.has("payload")) {
+                forge = jsonRecord
+                        .getJSONObject("payload")
+                        .getString("forge".replaceAll("[\\n\\t ]", ""));
+            }
+
+            logger.info("forge = " + forge);
+
+            //TODO: what if forge = null? Throw an exception?
+
+            if (forge != null) {
+                utils.insertMetadataIntoDB(forge, jsonRecord);
```

Comment:
Also, I think that all inserts should be done within one transaction because otherwise, we might end up with partial data.
You can see how transaction management with retries works [here](https://github.com/fasten-project/fasten/blob/develop/analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabasePlugin.java#L140)

Reply:
I followed closely the implementation of the VulnerabilityConsumer - I do only updates of the metadata in the callables (using updateCallableMetadata from MetadataDao). Therefore, I have changed the function names to better illustrate this.

Reply:
As long as you are certain that we won't be left with partial data in the case when your plugin unexpectedly terminates in the middle of processing, I am fine with what you have. But if there is a chance that part of the data will be written and the other part won't, then proper transaction management should be implemented.

Reply:
OK, done, I refactored the suggested code. BTW, the code from MetadataDatabasePlugin should be refactored and simplified IMO.

Reply:
Thanks. Yes, it is already being refactored by Stefanos (AUEB) to allow the processing of C and Python callgraphs.